### DISCUSSION TITLE
brew install gsl not working

### DISCUSSION THREAD

**Discussion by j-churchill** [Homebrew/discussions#3141] [Everyday usage] [answered]
Hi there, I'm trying to install gsl on my Apple Silicon M1 using brew and the documentation says to type "brew install gsl". When I do this, it says: 
`Warning: gsl 2.7.1 is already installed and up-to-date.
To reinstall 2.7.1, run:
  brew reinstall gsl`
and appears to be installed successfully, however, when I then try to compile c++ code using `g++ -Wall -o test.out test.cpp -lgslcblas` is says: 
`fatal error: gsl/gsl_math.h: No such file or directory
   13 | #include <gsl/gsl_math.h> |
compilation terminated.`

Any suggestions as to how to link gsl to the c++ compiler?

**Comment by j-churchill** [comment, marked as answer]
Solution from @rouzbehy:
g++ -Wall -I/opt/homebrew/Cellar/gsl/2.7.1/include -c test.cpp -o test.o 
g++ -L/opt/homebrew/Cellar/gsl/2.7.1/lib test.o -lgsl -lgslcblas -o test.out

**Comment by carlocab** [comment]
Try passing `-I/opt/homebrew/include` and `-L/opt/homebrew/lib` to the compiler.

**Comment by carlocab** [comment]
> Solution from @rouzbehy:
> 
> g++ -Wall -I/opt/homebrew/Cellar/gsl/2.7.1/include -c test.cpp -o test.o 
> 
> g++ -L/opt/homebrew/Cellar/gsl/2.7.1/lib test.o -lgsl -lgslcblas -o test.out

You should avoid referring to `Cellar` paths. Those are unstable and could change at any time. Instead, use

```
g++ -Wall -I/opt/homebrew/opt/gsl/include -c test.cpp -o test.o 
```
or
```
g++ -L/opt/homebrew/opt/gsl/lib test.o -lgsl -lgslcblas -o test.out
```
Alternatively, these should work too as long as you've never done `brew unlink gsl`:

```
g++ -Wall -I/opt/homebrew/include -c test.cpp -o test.o 
```
```
g++ -L/opt/homebrew/lib test.o -lgsl -lgslcblas -o test.out
```

**Reply by j-churchill** [reply]
Okay great, thanks!
